### PR TITLE
Fuse grad division into async grad allreduce

### PR DESCRIPTION
**Pull request by xrennvidia**
# What does this PR do ?

Fuse gradient division into torch.distributed.allreduce(). 

# Changelog 
This change is made for async grad AR in class MainParamsOptimizerWrappe.

# Usage
**This fusion requires at least PyTorch 22.04.**
This fusion is enabled by default, you can turn it off by setting grad_div_ar_fusion=False

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/core/optim/optimizer_with_main_params.py [hunk 1]
```diff
@@ -187,13 +188,16 @@ def __init__(
         # used with tensor parallel only (no pipeline parallelism)
         # be careful, weight update cannot start until all async grad AR works are done
         self._async_grad_allreduce = async_grad_allreduce
+        self._grad_divisor = 1 / float(get_data_parallel_world_size())
```
Comment (MaximumEntropy) [inline nemo/core/optim/optimizer_with_main_params.py:191]: We probably don't need `float()` here? python 3 div should be float anyway right?

**Review by MaximumEntropy — COMMENTED** [review]
Minor comment

File: nemo/core/optim/optimizer_with_main_params.py (hunk 1 above)
Comment (xrennvidia) [reply nemo/core/optim/optimizer_with_main_params.py]: Oh, yes. This seems redundant. Let me fix it. Thanks.

Comment (xrennvidia) [reply nemo/core/optim/optimizer_with_main_params.py]: fixed. check it out.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**